packages/python/plotly/plotly/express/_core.py

```diff
@@ -1307,37 +1307,36 @@ def build_dataframe(args, constructor):
 
     # Cast data_frame argument to DataFrame (it could be a numpy array, dict etc.)
     df_provided = args["data_frame"] is not None
+    needs_interchanging = False
     if df_provided and not isinstance(args["data_frame"], pd.DataFrame):
         if hasattr(args["data_frame"], "__dataframe__") and version.parse(
             pd.__version__
         ) >= version.parse("2.0.2"):
             import pandas.api.interchange
 
             df_not_pandas = args["data_frame"]
-            try:
-                df_pandas = pandas.api.interchange.from_dataframe(df_not_pandas)
-            except (ImportError, NotImplementedError) as exc:
-                # temporary workaround; developers of third-party libraries themselves
-                # should try a different implementation, if available. For example:
-                # def __dataframe__(self, ...):
-                #   if not some_condition:
-                #     self.to_pandas(...)
-                if not hasattr(df_not_pandas, "to_pandas"):
-                    raise exc
-                df_pandas = df_not_pandas.to_pandas()
-            args["data_frame"] = df_pandas
+            args["data_frame"] = df_not_pandas.__dataframe__()
+            columns = args["data_frame"].column_names()
```

Comment:
Can we be 100% sure anything returned by `__dataframe__()` will have `column_names` and `select_columns_by_name` methods? If there's any chance an object will come in with either of these missing we should fall back on interchanging the whole thing up front.

Reply:
yup! you can check the spec here: https://data-apis.org/dataframe-protocol/latest/API.html

Reply:
Oh I know they're in the spec, but I also know not everyone follows a spec to the letter 😉

Reply:
this can help highlight shortcomings in their implementation then 😉 I tried it out with polars and it works fine there

Reply:
I guess we'll know how to respond when we see:
`AttributeError: 'MyDataFrame' object has no attribute 'select_columns_by_name'`
And in principle you're right that it's not our problem, but we'll be the ones responding to the issue and having to tell our users "don't use this dataframe directly until they fix it." Whereas if we caught this case explicitly we could emit a warning like "This dataframe only partially implements the dataframe interchange protocol. Falling back on a slower full-copy algorithm" so it wouldn't affect usage in px, only performance, and it would be clear where the issue needs to be raised.

Reply:
thanks for explaining - OK I've added a condition so it'll only use `select_columns_by_name` if that attribute is present

Reply:
Nice. I took a look at also adding a fallback for missing `column_names` and that would be pretty awkward... but if someone has a partial implementation of the protocol presumably `column_names` is an easy piece so would get included early, whereas `select_columns_by_name` could be trickier. So let's leave it as you have it now. Thanks!

Reply:
yeah if they don't have `column_names` then `from_dataframe` wouldn't work either, as it uses that internally

https://github.com/pandas-dev/pandas/blob/92792ec063031ae41443dabeb9d12f8daaac3ef1/pandas/core/interchange/from_dataframe.py#L112